sno/fast_import.py

```diff
@@ -104,8 +105,57 @@ def fast_import_tables(
                 raise ValueError(f"{blob_path} already exists")
 
         for source in sources:
+            replacing_dataset = None
             if replace_existing == ReplaceExisting.GIVEN:
                 p.stdin.write(f"D {source.dest_path}\n".encode('utf8'))
+
+                # Optimisation: Try to avoid rewriting features for compatible schema changes.
+                # i.e. we can avoid rewriting feature blobs if the only schema changes are:
+                #  * dropped columns
+                #  * renamed columns
+                #  * reordered columns
+                try:
+                    replacing_dataset = RepositoryStructure(repo)[source.dest_path]
+                except KeyError:
+                    pass
+                else:
+                    # If the schemas are *identical, there's no need to pass this dataset object on.
+                    # Features won't be rewritten anyway.
+                    if replacing_dataset.schema == source.schema:
+                        replacing_dataset = None
+                    else:
+                        # Check that the schemas are compatible.
+                        new_col_ids = {col.id for col in source.schema}
+                        old_col_ids = {col.id for col in replacing_dataset.schema}
+                        new_types = {col.id: col.data_type for col in source.schema}
+                        old_types = {
+                            col.id: col.data_type
+                            for col in replacing_dataset.schema
+                            if col.id in new_col_ids
+                        }
+                        if new_col_ids - old_col_ids or new_types != old_types:
+                            # a column has been added, or a column's data_type has changed.
+                            # In this case all rows will be rewritten; can't avoid that.
+                            replacing_dataset = None
```

Comment:
There is actually only one schema change that guarantees that we need to rewrite all the data - changes to which column is the primary key. These probably don't work so great right now anyway, but you can test for them with Schema.is_pk_compatible.
- major type changes are supposed to be impossible - you can't just change a column from geometry to int
- minor type changes don't mean we need to rewrite values - if someone changes a column from int32 to int64, we don't need to rewrite all the ints. These minor types are just to help us generate working copies of the appropriate type
- column adds don't necessarily mean that we need to rewrite all the data either. Rewrites are only required for those rows which have non-null values in the new column - if a huge dataset with many rows has a new column added, it is highly likely that the user won't immediately have any data for the new column for the majority of existing rows, only for a few new rows.

Reply:
ok so basically as long as the PK fields don't change, I'll use `replacing_dataset`